core/src/main/java/io/parsingdata/metal/expression/value/reference/CurrentIteration.java

```diff
@@ -36,12 +38,33 @@
  */
 public class CurrentIteration implements ValueExpression {
 
+    private final ValueExpression level;
+
+    public CurrentIteration(final ValueExpression level) {
+        this.level = level;
+    }
+
     @Override
     public ImmutableList<Optional<Value>> eval(final ParseState parseState, final Encoding encoding) {
-        if (parseState.iterations.head == null) {
+        int level = getLevel(parseState, encoding);
+        ImmutableList<BigInteger> iterations = parseState.iterations;
+        while (level > 0 && iterations.tail != null) {
+            iterations = iterations.tail;
+            level--;
+        }
+
+        if (iterations.head == null || level > 0) {
             return ImmutableList.create(Optional.empty());
         }
-        return ImmutableList.create(Optional.of(createFromNumeric(parseState.iterations.head, new Encoding())));
+        return ImmutableList.create(Optional.of(createFromNumeric(iterations.head, new Encoding())));
+    }
+
+    private int getLevel(final ParseState parseState, final Encoding encoding) {
+        ImmutableList<Optional<Value>> evaluatedLevel = level.eval(parseState, encoding);
+        if (evaluatedLevel.size != 1 || !evaluatedLevel.head.isPresent()) {
+            throw new IllegalArgumentException("Level must evaluate to a single non-empty value.");
+        }
+        return evaluatedLevel.head.get().asNumeric().intValueExact();
```

Comment:
Metal uses `BigInteger` wherever possible, I don't think you need to revert to using an `int` here.

Reply:
Agreed. Using BigInteger instead.